From 7644ddc0e80eb9aade37b2218afc3be7826d6153 Mon Sep 17 00:00:00 2001
From: Xianzhong <b07117@freescale.com>
Date: Sat, 1 Sep 2012 21:12:51 +0800
Subject: [PATCH 216/280] ENGR00221688 Fixed video random hang and mosaic
 issues

Vpu has the implicit page alignment requirement for buffer address,
Has to round it to page edge in gpu memory allocator

And add trace message for gpu memory free

Signed-off-by: Xianzhong <b07117@freescale.com>
---
 vpu/vpu_io.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/vpu/vpu_io.c b/vpu/vpu_io.c
index 2fba7cf..e58a2ef 100644
--- a/vpu/vpu_io.c
+++ b/vpu/vpu_io.c
@@ -417,7 +417,7 @@ error:
 	ion_close(fd);
 	return ret;
 #elif USE_GPU
-        bytes = buff->size;
+        bytes = buff->size + PAGE_SIZE;
         phy_addr = 0; virt_addr = NULL;
         ret = gcoOS_AllocateVideoMemory(NULL, 1, 0, &bytes, &phy_addr,&virt_addr,&handle);
         if(ret != 0) {
@@ -428,6 +428,11 @@ error:
         buff->virt_uaddr = (unsigned long)virt_addr;
         buff->phy_addr = (unsigned long)phy_addr;
         buff->cpu_addr = (unsigned long)handle;
+
+        //vpu requires page alignment for the address implicitly, round it to page edge
+        buff->virt_uaddr = (buff->virt_uaddr + PAGE_SIZE -1) & ~(PAGE_SIZE -1);
+        buff->phy_addr = (buff->phy_addr + PAGE_SIZE -1) & ~(PAGE_SIZE -1);
+
         info_msg("<gpu> alloc handle: 0x%x, paddr: 0x%x, vaddr: 0x%x",
 			(unsigned int)handle, (unsigned int)buff->phy_addr,
 			(unsigned int)buff->virt_uaddr);
@@ -547,6 +552,10 @@ int _IOFreePhyMem(int which, vpu_mem_desc * buff)
            err_msg("%d, gpu allocator failed to free handle 0x%x", ret, (unsigned int)handle);
            return ret;
         }
+
+        info_msg("<gpu> free handle: 0x%x, paddr: 0x%x, vaddr: 0x%x",
+			(unsigned int)handle, (unsigned int)buff->phy_addr,
+			(unsigned int)buff->virt_uaddr);
 #else
 	int fd_pmem;
 
-- 
1.8.0

